lib/ocpp/common/websocket/websocket_libwebsockets.cpp

```diff
@@ -675,13 +677,23 @@ bool WebsocketTlsTPM::connect() {
 
     if (!connected) {
         EVLOG_error << "Conn failed, interrupting.";
-        // Interrupt and drop the connection data
-        std::shared_ptr<ConnectionData> local_data = conn_data;
-        if (local_data != nullptr) {
-            local_data->do_interrupt();
+
+        // If we timeouted the on_conn_fail was not dispatched, since it did not had the chance
+        if (timeouted && local_data->get_state() != EConnectionState::ERROR) {
+            EVLOG_error << "Conn failed with timeout, without disconnect dispatch, dispatching manually.";
+            on_conn_fail();
         }
 
-        conn_data.reset();
+        // Interrupt and drop the connection data
+        local_data->do_interrupt();
+
+        // Also interrupt the latest conenction, if it was set by a parallel thread
+        auto local = conn_data;
+
+        if (local != nullptr) {
+            local->do_interrupt();
+            conn_data.reset();
```

Comment:
should conn_data always be reset?

Reply:
if we checked and it is always null, whould we re-set an already null null pointer? During the execution of this, I don't think that any thread should run in paralel that can somehow set the stat of the conn_data in any way, Have you detected any possibility of that in code?

Reply:
With multiple threads there is always a possibility of a race condition unless critical sections are protected by mutexes.
e.g.
thread 1: line  961 auto local = conn_data; and then gets swapped out
thread 2: calls connect() ... and updates conn_data line 609 conn_data = local_data; and then gets swapped out

There is a comment in the code that connect() "Will be called from external threads as well"